# ğŸŒ Web API Reference

**ë¬¸ì„œ ë²„ì „**: 1.0
**ìµœì¢… ì—…ë°ì´íŠ¸**: 2025-12-15
**Base URL**: `http://localhost:8000`

---

## ëª©ì°¨

1. [ê°œìš”](#1-ê°œìš”)
2. [ì¸ì¦ ë° ë³´ì•ˆ](#2-ì¸ì¦-ë°-ë³´ì•ˆ)
3. [Endpoints](#3-endpoints)
   - [3.1 Health Check](#31-health-check)
   - [3.2 Web UI](#32-web-ui)
   - [3.3 Inspection Endpoints](#33-inspection-endpoints)
   - [3.4 Batch Processing](#34-batch-processing)
   - [3.5 Comparison & Analytics](#35-comparison--analytics)
   - [3.6 Results Management](#36-results-management)
4. [ë°ì´í„° ìŠ¤í‚¤ë§ˆ](#4-ë°ì´í„°-ìŠ¤í‚¤ë§ˆ)
5. [ì—ëŸ¬ ì²˜ë¦¬](#5-ì—ëŸ¬-ì²˜ë¦¬)
6. [ì˜ˆì œ ì½”ë“œ](#6-ì˜ˆì œ-ì½”ë“œ)

---

## 1. ê°œìš”

Color Meter Web APIëŠ” **FastAPI** ê¸°ë°˜ì˜ RESTful APIë¡œ, ì½˜íƒíŠ¸ë Œì¦ˆ ìƒ‰ìƒ ê²€ì‚¬ë¥¼ ìœ„í•œ í”„ë¡œê·¸ë˜ë§¤í‹± ì¸í„°í˜ì´ìŠ¤ë¥¼ ì œê³µí•©ë‹ˆë‹¤.

**ì£¼ìš” ê¸°ëŠ¥**:
- âœ… ë‹¨ì¼/ë°°ì¹˜ ì´ë¯¸ì§€ ê²€ì‚¬
- âœ… íŒŒë¼ë¯¸í„° ì¬ê³„ì‚° (ì´ë¯¸ì§€ ì¬ì—…ë¡œë“œ ë¶ˆí•„ìš”)
- âœ… ë¡œíŠ¸ ë¹„êµ ë¶„ì„ (Golden Sample vs Test Images)
- âœ… ê²°ê³¼ ì €ì¥ ë° ì¡°íšŒ
- âœ… ì‰í¬ ë¶„ì„ (Zone-based + Image-based)

**ê¸°ìˆ  ìŠ¤íƒ**:
- FastAPI 0.124+
- Uvicorn (ASGI Server)
- Python 3.8+

**Swagger UI**: `http://localhost:8000/docs`
**ReDoc**: `http://localhost:8000/redoc`

---

## 2. ì¸ì¦ ë° ë³´ì•ˆ

### 2.1 í˜„ì¬ ìƒíƒœ

**ì¸ì¦**: ì—†ìŒ (ë¡œì»¬ ê°œë°œ í™˜ê²½)

**ë³´ì•ˆ ê³ ë ¤ì‚¬í•­**:
- í”„ë¡œë•ì…˜ ë°°í¬ ì‹œ API Key ë˜ëŠ” OAuth2 ì¸ì¦ ê¶Œì¥
- CORS ì„¤ì •: í˜„ì¬ ëª¨ë“  origin í—ˆìš© (`allow_origins=["*"]`)
- íŒŒì¼ ì—…ë¡œë“œ ì œí•œ: ìµœëŒ€ 50MB

### 2.2 í–¥í›„ ê³„íš

**Phase 4 (ì˜ˆì •)**:
```python
# API Key ì¸ì¦
headers = {"X-API-Key": "your-api-key"}
response = requests.post(url, headers=headers, ...)
```

---

## 3. Endpoints

### 3.1 Health Check

#### `GET /health`

ì„œë²„ ìƒíƒœ í™•ì¸

**Request**:
```http
GET /health HTTP/1.1
Host: localhost:8000
```

**Response** (200 OK):
```json
{
  "status": "ok",
  "timestamp": "2025-12-15T10:30:00Z"
}
```

**ì‚¬ìš© ì‚¬ë¡€**:
- ì„œë²„ ê°€ë™ ì—¬ë¶€ í™•ì¸
- í—¬ìŠ¤ ì²´í¬ ëª¨ë‹ˆí„°ë§
- Docker ì»¨í…Œì´ë„ˆ readiness probe

---

### 3.2 Web UI

#### `GET /`

Web UI ë©”ì¸ í˜ì´ì§€ (HTML)

**Request**:
```http
GET / HTTP/1.1
Host: localhost:8000
```

**Response** (200 OK):
- Content-Type: `text/html`
- ì¸í„°ë™í‹°ë¸Œ ì›¹ ì¸í„°í˜ì´ìŠ¤

**ê¸°ëŠ¥**:
- ì´ë¯¸ì§€ ì—…ë¡œë“œ ë° ì‹œê°í™”
- ì‹¤ì‹œê°„ íŒŒë¼ë¯¸í„° ì¡°ì •
- ê²°ê³¼ ì‹œê°í™” (6ê°œ íƒ­)

---

### 3.3 Inspection Endpoints

#### `POST /inspect`

ë‹¨ì¼ ì´ë¯¸ì§€ ê²€ì‚¬

**Request**:
```http
POST /inspect HTTP/1.1
Host: localhost:8000
Content-Type: multipart/form-data

--boundary
Content-Disposition: form-data; name="file"; filename="lens.jpg"
Content-Type: image/jpeg

[ì´ë¯¸ì§€ ë°”ì´ë„ˆë¦¬ ë°ì´í„°]
--boundary
Content-Disposition: form-data; name="sku"

SKU001
--boundary
Content-Disposition: form-data; name="run_judgment"

true
--boundary--
```

**Parameters**:
| ì´ë¦„ | íƒ€ì… | í•„ìˆ˜ | ì„¤ëª… |
|------|------|------|------|
| `file` | File | âœ… | ê²€ì‚¬í•  ì´ë¯¸ì§€ (JPG, PNG) |
| `sku` | string | âœ… | SKU ì½”ë“œ (ì˜ˆ: `SKU001`) |
| `run_judgment` | boolean | âŒ | OK/NG íŒì • ì‹¤í–‰ ì—¬ë¶€ (ê¸°ë³¸: `true`) |

**Response** (200 OK):
```json
{
  "status": "success",
  "image_id": "abc123def456",
  "session_id": "session_789xyz",
  "judgment": "OK",
  "overall_delta_e": 2.5,
  "confidence": 0.92,
  "zones": [
    {
      "zone_name": "Zone_C",
      "measured": {"L": 72.1, "a": 137.2, "b": 122.7},
      "expected": {"L": 72.2, "a": 137.3, "b": 122.8},
      "delta_e": 0.15,
      "threshold": 4.0,
      "passed": true,
      "std_L": 5.2,
      "pixel_count": 35200
    }
  ],
  "decision_trace": {
    "final": "OK",
    "because": "All zones passed",
    "zone_checks": {...}
  },
  "next_actions": ["ì¶œí•˜ ìŠ¹ì¸"],
  "confidence_breakdown": {
    "pixel_count_score": 0.95,
    "transition_score": 0.90,
    "std_score": 0.92,
    "sector_uniformity": 0.88,
    "lens_detection": 0.95,
    "overall": 0.92
  },
  "ink_analysis": {
    "zone_based": {
      "ink_count": 1,
      "inks": [{
        "zone_name": "Zone_C",
        "measured": {"L": 72.1, "a": 137.2, "b": 122.7},
        "rgb": [195, 150, 135],
        "hex": "#C39687"
      }]
    },
    "image_based": {
      "ink_count": 1,
      "inks": [{
        "L": 71.8, "a": 137.0, "b": 122.5,
        "weight": 1.0,
        "rgb": [194, 149, 134],
        "hex": "#C29586"
      }],
      "meta": {
        "bic": -125430.5,
        "correction_applied": false,
        "sampling_config": {
          "chroma_threshold": 6.0,
          "sampled_pixels": 45200
        }
      }
    }
  },
  "profile_data": {...},
  "overlay_url": "/results/session_789xyz/overlay.png"
}
```

**Error Responses**:
- `400 Bad Request`: íŒŒì¼ ëˆ„ë½, SKU ì—†ìŒ
- `404 Not Found`: SKU íŒŒì¼ ì—†ìŒ
- `500 Internal Server Error`: ê²€ì‚¬ íŒŒì´í”„ë¼ì¸ ì—ëŸ¬

---

#### `POST /inspect_v2`

í™•ì¥ëœ ê²€ì‚¬ ì—”ë“œí¬ì¸íŠ¸ (ì¶”ê°€ ì§„ë‹¨ ì •ë³´)

**ì°¨ì´ì **:
- `risk_factors` í¬í•¨
- `uniformity_metrics` ìƒì„¸ ì •ë³´
- `sector_statistics` í¬í•¨

**Request**: `/inspect`ì™€ ë™ì¼

**Response** (200 OK):
```json
{
  // ... /inspect ì‘ë‹µ ë‚´ìš© ...
  "risk_factors": [
    {
      "category": "sector_uniformity",
      "severity": "medium",
      "message": "Zone B ì„¹í„° ê°„ í¸ì°¨ ë³´í†µ",
      "details": {
        "zone": "B",
        "max_sector_std_L": 6.5,
        "worst_sector": 3
      }
    }
  ],
  "uniformity_metrics": {
    "max_std_L": 8.5,
    "warning_threshold": 10.0,
    "retake_threshold": 12.0
  },
  "sector_statistics": {
    "enabled": true,
    "num_sectors": 8,
    "max_sector_std_L": 6.5,
    "worst_zone": "B"
  }
}
```

---

#### `POST /recompute`

íŒŒë¼ë¯¸í„° ì¬ê³„ì‚° (ì´ë¯¸ì§€ ì¬ì—…ë¡œë“œ ë¶ˆí•„ìš”) â­ PHASE7 ì‹ ê·œ

**ëª©ì **:
- ê°™ì€ ì´ë¯¸ì§€ì— ëŒ€í•´ ë‹¤ë¥¸ íŒŒë¼ë¯¸í„°ë¡œ ì¬ë¶„ì„
- íŒŒë¼ë¯¸í„° íŠœë‹ ì›Œí¬í”Œë¡œìš° ìµœì í™” (30Ã— ì†ë„ í–¥ìƒ)

**Request**:
```http
POST /recompute HTTP/1.1
Host: localhost:8000
Content-Type: application/x-www-form-urlencoded

image_id=abc123def456&sku=SKU001&params={"smoothing_window":15,"min_gradient":3.0}
```

**Parameters**:
| ì´ë¦„ | íƒ€ì… | í•„ìˆ˜ | ì„¤ëª… |
|------|------|------|------|
| `image_id` | string | âœ… | `/inspect` ì‘ë‹µì˜ `image_id` |
| `sku` | string | âœ… | SKU ì½”ë“œ |
| `params` | JSON string | âœ… | ì¡°ì •í•  íŒŒë¼ë¯¸í„° (ì•„ë˜ ì°¸ì¡°) |
| `run_judgment` | boolean | âŒ | íŒì • ì‹¤í–‰ ì—¬ë¶€ |

**ì§€ì› íŒŒë¼ë¯¸í„°** (12ê°œ):
```json
{
  "detection_method": "gradient",        // gradient, delta_e, hybrid, variable_width
  "smoothing_window": 15,                // 1-100
  "min_gradient": 3.0,                   // 0.0-10.0
  "min_delta_e": 3.0,                    // 0.0-10.0
  "correction_method": "polynomial",     // none, polynomial, retinex, clahe, auto
  "correction_degree": 2,                // 1-4
  "zone_method": "auto",                 // auto, uniform
  "expected_zones": 1,                   // 1-10
  "uniformity_threshold_L": 10.0,        // 0.0-50.0
  "uniformity_threshold_a": 15.0,        // 0.0-50.0
  "uniformity_threshold_b": 15.0,        // 0.0-50.0
  "subpixel_refinement": true            // true/false
}
```

**Response** (200 OK):
- `/inspect`ì™€ ë™ì¼í•œ ì‘ë‹µ í˜•ì‹
- ìƒˆë¡œìš´ íŒŒë¼ë¯¸í„°ë¡œ ì¬ê³„ì‚°ëœ ê²°ê³¼

**Error Responses**:
- `404 Not Found`: `image_id`ê°€ ìºì‹œì— ì—†ìŒ (TTL 15ë¶„)
- `400 Bad Request`: ì˜ëª»ëœ íŒŒë¼ë¯¸í„°

**ì‚¬ìš© ì‚¬ë¡€**:
```python
# 1. ì´ë¯¸ì§€ ì—…ë¡œë“œ
resp1 = requests.post("/inspect", files={"file": img}, data={"sku": "SKU001"})
image_id = resp1.json()["image_id"]

# 2. íŒŒë¼ë¯¸í„° íŠœë‹ (ì¦‰ì‹œ ì‹¤í–‰)
resp2 = requests.post("/recompute", data={
    "image_id": image_id,
    "sku": "SKU001",
    "params": json.dumps({"smoothing_window": 20, "min_gradient": 2.5})
})

# 3. ë‹¤ë¥¸ íŒŒë¼ë¯¸í„°ë¡œ ë‹¤ì‹œ í…ŒìŠ¤íŠ¸
resp3 = requests.post("/recompute", data={
    "image_id": image_id,
    "sku": "SKU001",
    "params": json.dumps({"expected_zones": 2})
})
```

---

### 3.4 Batch Processing

#### `POST /batch`

ë°°ì¹˜ ì´ë¯¸ì§€ ê²€ì‚¬

**Request (ZIP íŒŒì¼)**:
```http
POST /batch HTTP/1.1
Host: localhost:8000
Content-Type: multipart/form-data

--boundary
Content-Disposition: form-data; name="zip_file"; filename="batch.zip"
Content-Type: application/zip

[ZIP ë°”ì´ë„ˆë¦¬ ë°ì´í„°]
--boundary
Content-Disposition: form-data; name="sku"

SKU001
--boundary--
```

**Request (ì„œë²„ ê²½ë¡œ)**:
```http
POST /batch HTTP/1.1
Host: localhost:8000
Content-Type: application/json

{
  "path": "/data/raw_images/batch_001/",
  "sku": "SKU001",
  "run_judgment": true
}
```

**Parameters**:
| ì´ë¦„ | íƒ€ì… | í•„ìˆ˜ | ì„¤ëª… |
|------|------|------|------|
| `zip_file` | File | âŒ | ZIP íŒŒì¼ (ë˜ëŠ” `path` ì‚¬ìš©) |
| `path` | string | âŒ | ì„œë²„ ìƒì˜ ì´ë¯¸ì§€ í´ë” ê²½ë¡œ |
| `sku` | string | âœ… | SKU ì½”ë“œ |
| `run_judgment` | boolean | âŒ | íŒì • ì‹¤í–‰ ì—¬ë¶€ |

**Response** (200 OK):
```json
{
  "status": "success",
  "run_id": "batch_20251215_103045",
  "total_images": 50,
  "processed": 50,
  "failed": 0,
  "summary": {
    "ok_count": 42,
    "ng_count": 5,
    "retake_count": 3,
    "ok_with_warning_count": 0,
    "avg_delta_e": 2.8,
    "max_delta_e": 5.3,
    "avg_confidence": 0.87
  },
  "results": [
    {
      "filename": "lens_001.jpg",
      "judgment": "OK",
      "overall_delta_e": 2.1,
      "confidence": 0.92
    },
    {
      "filename": "lens_002.jpg",
      "judgment": "NG",
      "overall_delta_e": 5.3,
      "confidence": 0.88,
      "ng_reasons": ["Zone_A"]
    }
  ],
  "csv_url": "/results/batch_20251215_103045/summary.csv"
}
```

**Error Responses**:
- `400 Bad Request`: ZIP íŒŒì¼ê³¼ ê²½ë¡œ ëª¨ë‘ ëˆ„ë½
- `404 Not Found`: ê²½ë¡œê°€ ì¡´ì¬í•˜ì§€ ì•ŠìŒ

---

### 3.5 Comparison & Analytics

#### `POST /compare`

ë¡œíŠ¸ ë¹„êµ ë¶„ì„ (Golden Sample vs Test Images) â­ PHASE7 ì‹ ê·œ

**ëª©ì **:
- ê¸°ì¤€ ì´ë¯¸ì§€(Golden Sample) ëŒ€ë¹„ í…ŒìŠ¤íŠ¸ ì´ë¯¸ì§€ë“¤ì˜ ìƒ‰ìƒ í¸ì°¨ ë¶„ì„
- ë¡œíŠ¸ ì „ì²´ì˜ ì¼ê´€ì„± í‰ê°€
- ì´ìƒì¹˜(Outlier) ê°ì§€

**Request**:
```http
POST /compare HTTP/1.1
Host: localhost:8000
Content-Type: multipart/form-data

--boundary
Content-Disposition: form-data; name="reference_file"; filename="golden.jpg"
Content-Type: image/jpeg

[ê¸°ì¤€ ì´ë¯¸ì§€]
--boundary
Content-Disposition: form-data; name="test_files"; filename="lot_001.jpg"
Content-Type: image/jpeg

[í…ŒìŠ¤íŠ¸ ì´ë¯¸ì§€ 1]
--boundary
Content-Disposition: form-data; name="test_files"; filename="lot_002.jpg"
Content-Type: image/jpeg

[í…ŒìŠ¤íŠ¸ ì´ë¯¸ì§€ 2]
--boundary
Content-Disposition: form-data; name="sku"

SKU001
--boundary--
```

**Parameters**:
| ì´ë¦„ | íƒ€ì… | í•„ìˆ˜ | ì„¤ëª… |
|------|------|------|------|
| `reference_file` | File | âœ… | ê¸°ì¤€ ì´ë¯¸ì§€ (Golden Sample) |
| `test_files` | File[] | âœ… | í…ŒìŠ¤íŠ¸ ì´ë¯¸ì§€ ë°°ì—´ (1~100ê°œ) |
| `sku` | string | âœ… | SKU ì½”ë“œ |

**Response** (200 OK):
```json
{
  "status": "success",
  "reference": {
    "filename": "golden.jpg",
    "zones": [
      {
        "name": "Zone_C",
        "mean_L": 72.2,
        "mean_a": 137.3,
        "mean_b": 122.8
      }
    ]
  },
  "tests": [
    {
      "filename": "lot_001.jpg",
      "zone_deltas": [
        {
          "zone": "Zone_C",
          "delta_L": -0.5,
          "delta_a": 0.3,
          "delta_b": 0.2,
          "delta_e": 0.6
        }
      ],
      "overall_shift": "Slightly darker",
      "max_delta_e": 0.6
    },
    {
      "filename": "lot_002.jpg",
      "zone_deltas": [
        {
          "zone": "Zone_C",
          "delta_L": -2.3,
          "delta_a": 1.5,
          "delta_b": 0.8,
          "delta_e": 2.8
        }
      ],
      "overall_shift": "Darker and more yellow",
      "max_delta_e": 2.8
    }
  ],
  "batch_summary": {
    "mean_delta_e_per_zone": {
      "Zone_C": 1.7
    },
    "max_delta_e_per_zone": {
      "Zone_C": 2.8
    },
    "std_delta_e_per_zone": {
      "Zone_C": 1.1
    },
    "stability_score": 0.83,
    "outliers": ["lot_002.jpg"]
  }
}
```

**Response Fields**:
| í•„ë“œ | ì„¤ëª… |
|------|------|
| `overall_shift` | ìƒ‰ìƒ ë³€í™” ë°©í–¥ (Darker, Lighter, More red, etc.) |
| `stability_score` | ë¡œíŠ¸ ì¼ê´€ì„± ì ìˆ˜ (0.0~1.0, ë†’ì„ìˆ˜ë¡ ì¢‹ìŒ) |
| `outliers` | ì´ìƒì¹˜ë¡œ íŒë‹¨ëœ ì´ë¯¸ì§€ íŒŒì¼ëª… ë°°ì—´ |

**Stability Score ê³„ì‚°**:
```python
stability_score = 1.0 - min(mean_delta_e / 10.0, 1.0)
```

**Outlier ê¸°ì¤€**:
```python
threshold = mean + 2.0 * std
is_outlier = (delta_e > threshold)
```

**ì‚¬ìš© ì‚¬ë¡€**:
```python
# ë¡œíŠ¸ QC (Quality Control)
response = requests.post("/compare", files={
    "reference_file": open("golden_sample.jpg", "rb"),
    "test_files": [
        open("lot_001_sample1.jpg", "rb"),
        open("lot_001_sample2.jpg", "rb"),
        # ... ìµœëŒ€ 100ê°œ
    ]
}, data={"sku": "SKU001"})

# ê²°ê³¼ ë¶„ì„
summary = response.json()["batch_summary"]
if summary["stability_score"] < 0.7:
    print("âš ï¸ ë¡œíŠ¸ ì¼ê´€ì„± ë‚®ìŒ")

outliers = summary["outliers"]
if outliers:
    print(f"ì´ìƒì¹˜ ê°ì§€: {outliers}")
```

---

### 3.6 Results Management

#### `GET /results/{run_id}`

ë°°ì¹˜ ê²€ì‚¬ ê²°ê³¼ ì¡°íšŒ

**Request**:
```http
GET /results/batch_20251215_103045 HTTP/1.1
Host: localhost:8000
```

**Response** (200 OK):
```json
{
  "run_id": "batch_20251215_103045",
  "timestamp": "2025-12-15T10:30:45Z",
  "sku": "SKU001",
  "total_images": 50,
  "summary": {...},
  "results": [...]
}
```

**Error Responses**:
- `404 Not Found`: `run_id`ê°€ ì¡´ì¬í•˜ì§€ ì•ŠìŒ

---

#### `GET /results/{run_id}/{filename}`

ê°œë³„ ê²°ê³¼ íŒŒì¼ ë‹¤ìš´ë¡œë“œ

**Request**:
```http
GET /results/batch_20251215_103045/summary.csv HTTP/1.1
Host: localhost:8000
```

**Response** (200 OK):
- Content-Type: `text/csv` ë˜ëŠ” `image/png`
- íŒŒì¼ ë‹¤ìš´ë¡œë“œ

**ì§€ì› íŒŒì¼**:
- `summary.csv`: ë°°ì¹˜ ìš”ì•½ CSV
- `lens_001_result.json`: ê°œë³„ ê²°ê³¼ JSON
- `lens_001_overlay.png`: ì˜¤ë²„ë ˆì´ ì´ë¯¸ì§€

---

## 4. ë°ì´í„° ìŠ¤í‚¤ë§ˆ

### 4.1 InspectionResult

```typescript
interface InspectionResult {
  status: "success" | "error";
  image_id?: string;              // ì´ë¯¸ì§€ ìºì‹œ ID (ì¬ê³„ì‚°ìš©)
  session_id?: string;             // ì„¸ì…˜ ID
  judgment: "OK" | "OK_WITH_WARNING" | "NG" | "RETAKE";
  overall_delta_e: number;
  confidence: number;              // 0.0~1.0
  zones: ZoneResult[];
  decision_trace: DecisionTrace;
  next_actions: string[];
  confidence_breakdown: ConfidenceBreakdown;
  ink_analysis: InkAnalysis;
  profile_data?: ProfileData;
  overlay_url?: string;
  risk_factors?: RiskFactor[];     // /inspect_v2ì—ë§Œ
}
```

### 4.2 ZoneResult

```typescript
interface ZoneResult {
  zone_name: string;               // "Zone_C", "Zone_B", etc.
  measured: LabColor;
  expected: LabColor;
  delta_e: number;
  threshold: number;
  passed: boolean;
  std_L: number;                   // ê· ì¼ë„ (í‘œì¤€í¸ì°¨)
  std_a?: number;
  std_b?: number;
  pixel_count: number;
  rgb?: [number, number, number];
  hex?: string;
}
```

### 4.3 InkAnalysis

```typescript
interface InkAnalysis {
  zone_based: {
    ink_count: number;
    inks: ZoneBasedInk[];
  };
  image_based: {
    ink_count: number;
    inks: ImageBasedInk[];
    meta: {
      bic: number;
      correction_applied: boolean;
      sampling_config: SamplingConfig;
    };
  };
}

interface ZoneBasedInk {
  zone_name: string;
  measured: LabColor;
  rgb: [number, number, number];
  hex: string;
}

interface ImageBasedInk {
  L: number;
  a: number;
  b: number;
  weight: number;                  // í”½ì…€ ë¹„ìœ¨ (0.0~1.0)
  rgb: [number, number, number];
  hex: string;
}

interface SamplingConfig {
  chroma_threshold: number;
  L_max: number;
  highlight_removed: boolean;
  candidate_pixels: number;
  sampled_pixels: number;
  sampling_ratio: number;          // 0.0~1.0
}
```

### 4.4 DecisionTrace

```typescript
interface DecisionTrace {
  final: "OK" | "OK_WITH_WARNING" | "NG" | "RETAKE";
  because: string;                 // íŒì • ì´ìœ 
  overrides?: string;              // ì˜¤ë²„ë¼ì´ë“œ ì •ë³´
  zone_checks: {
    [zone: string]: {
      delta_e: number;
      threshold: number;
      passed: boolean;
    };
  };
  uniformity_check?: {
    max_std_l: number;
    warning_threshold: number;
    retake_threshold: number;
    status: "ok" | "warning" | "retake";
  };
}
```

### 4.5 ConfidenceBreakdown

```typescript
interface ConfidenceBreakdown {
  pixel_count_score: number;       // 0.0~1.0
  transition_score: number;
  std_score: number;
  sector_uniformity: number;
  lens_detection: number;
  overall: number;
}
```

### 4.6 RiskFactor

```typescript
interface RiskFactor {
  category: "delta_e_exceeded" | "sector_uniformity" | "uniformity_low" | "boundary_unclear" | "coverage_low";
  severity: "low" | "medium" | "high";
  message: string;
  details: {
    [key: string]: any;
  };
}
```

---

## 5. ì—ëŸ¬ ì²˜ë¦¬

### 5.1 ì—ëŸ¬ ì‘ë‹µ í˜•ì‹

```json
{
  "status": "error",
  "error_code": "SKU_NOT_FOUND",
  "message": "SKU 'SKU999' not found in database",
  "details": {
    "sku": "SKU999",
    "available_skus": ["SKU001", "SKU002"]
  }
}
```

### 5.2 ì—ëŸ¬ ì½”ë“œ ëª©ë¡

| ì½”ë“œ | HTTP Status | ì„¤ëª… |
|------|-------------|------|
| `FILE_NOT_PROVIDED` | 400 | íŒŒì¼ì´ ì—…ë¡œë“œë˜ì§€ ì•ŠìŒ |
| `SKU_NOT_PROVIDED` | 400 | SKU íŒŒë¼ë¯¸í„° ëˆ„ë½ |
| `INVALID_FILE_FORMAT` | 400 | ì§€ì›í•˜ì§€ ì•ŠëŠ” íŒŒì¼ í˜•ì‹ |
| `FILE_TOO_LARGE` | 400 | íŒŒì¼ í¬ê¸° ì´ˆê³¼ (>50MB) |
| `SKU_NOT_FOUND` | 404 | SKU íŒŒì¼ ì—†ìŒ |
| `IMAGE_ID_NOT_FOUND` | 404 | ìºì‹œëœ ì´ë¯¸ì§€ ì—†ìŒ (TTL ë§Œë£Œ) |
| `RUN_ID_NOT_FOUND` | 404 | ë°°ì¹˜ ê²°ê³¼ ì—†ìŒ |
| `PIPELINE_ERROR` | 500 | ê²€ì‚¬ íŒŒì´í”„ë¼ì¸ ì‹¤í–‰ ì‹¤íŒ¨ |
| `LENS_NOT_DETECTED` | 500 | ë Œì¦ˆ ê²€ì¶œ ì‹¤íŒ¨ |
| `ZONE_DETECTION_FAILED` | 500 | Zone ë¶„í•  ì‹¤íŒ¨ |

### 5.3 ì—ëŸ¬ ì²˜ë¦¬ ì˜ˆì œ

```python
response = requests.post("/inspect", files={...}, data={...})

if response.status_code != 200:
    error = response.json()
    error_code = error.get("error_code")

    if error_code == "SKU_NOT_FOUND":
        print(f"SKUë¥¼ ì°¾ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤: {error['details']['sku']}")
        print(f"ì‚¬ìš© ê°€ëŠ¥í•œ SKU: {error['details']['available_skus']}")
    elif error_code == "LENS_NOT_DETECTED":
        print("ë Œì¦ˆ ê²€ì¶œ ì‹¤íŒ¨. ì´ë¯¸ì§€ í’ˆì§ˆì„ í™•ì¸í•˜ì„¸ìš”.")
    else:
        print(f"ì—ëŸ¬ ë°œìƒ: {error['message']}")
```

---

## 6. ì˜ˆì œ ì½”ë“œ

### 6.1 Python (requests)

#### ë‹¨ì¼ ì´ë¯¸ì§€ ê²€ì‚¬

```python
import requests

url = "http://localhost:8000/inspect"

with open("lens_image.jpg", "rb") as f:
    response = requests.post(
        url,
        files={"file": f},
        data={"sku": "SKU001", "run_judgment": "true"}
    )

if response.status_code == 200:
    result = response.json()
    print(f"Judgment: {result['judgment']}")
    print(f"Î”E: {result['overall_delta_e']:.2f}")
    print(f"Confidence: {result['confidence']:.2f}")

    # ì‰í¬ ë¶„ì„
    zone_count = result["ink_analysis"]["zone_based"]["ink_count"]
    image_count = result["ink_analysis"]["image_based"]["ink_count"]
    print(f"Zone-based: {zone_count}ê°œ, Image-based: {image_count}ê°œ")

    if zone_count != image_count:
        print("âš ï¸ ì‰í¬ ê°œìˆ˜ ë¶ˆì¼ì¹˜ - SKU ì„¤ì • ê²€í†  í•„ìš”")
else:
    print(f"Error: {response.json()['message']}")
```

#### íŒŒë¼ë¯¸í„° ì¬ê³„ì‚°

```python
# 1. ì´ˆê¸° ê²€ì‚¬
resp1 = requests.post("/inspect", files={"file": img}, data={"sku": "SKU001"})
image_id = resp1.json()["image_id"]

# 2. íŒŒë¼ë¯¸í„° íŠœë‹
params = {
    "smoothing_window": 20,
    "min_gradient": 2.5,
    "expected_zones": 2
}

resp2 = requests.post("/recompute", data={
    "image_id": image_id,
    "sku": "SKU001",
    "params": json.dumps(params)
})

result2 = resp2.json()
print(f"ì¬ê³„ì‚° ê²°ê³¼: {result2['judgment']}")
```

#### ë°°ì¹˜ ê²€ì‚¬

```python
import zipfile
from pathlib import Path

# ZIP íŒŒì¼ ìƒì„±
with zipfile.ZipFile("batch.zip", "w") as zf:
    for img_path in Path("images/").glob("*.jpg"):
        zf.write(img_path, img_path.name)

# ë°°ì¹˜ ê²€ì‚¬
with open("batch.zip", "rb") as f:
    response = requests.post(
        "http://localhost:8000/batch",
        files={"zip_file": f},
        data={"sku": "SKU001"}
    )

result = response.json()
print(f"ì²˜ë¦¬: {result['processed']}/{result['total_images']}")
print(f"OK: {result['summary']['ok_count']}")
print(f"NG: {result['summary']['ng_count']}")
print(f"RETAKE: {result['summary']['retake_count']}")
```

#### ë¡œíŠ¸ ë¹„êµ

```python
files = {
    "reference_file": open("golden.jpg", "rb"),
    "test_files": [
        open("lot1.jpg", "rb"),
        open("lot2.jpg", "rb"),
        open("lot3.jpg", "rb")
    ]
}

response = requests.post(
    "http://localhost:8000/compare",
    files=files,
    data={"sku": "SKU001"}
)

result = response.json()
stability = result["batch_summary"]["stability_score"]

if stability < 0.7:
    print("âš ï¸ ë¡œíŠ¸ ì¼ê´€ì„± ë‚®ìŒ")
    outliers = result["batch_summary"]["outliers"]
    print(f"ì´ìƒì¹˜: {outliers}")
else:
    print(f"âœ… ë¡œíŠ¸ ì¼ê´€ì„± ì–‘í˜¸ ({stability:.2f})")
```

---

### 6.2 JavaScript (fetch)

```javascript
// ë‹¨ì¼ ì´ë¯¸ì§€ ê²€ì‚¬
async function inspectImage(file, sku) {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("sku", sku);
  formData.append("run_judgment", "true");

  const response = await fetch("http://localhost:8000/inspect", {
    method: "POST",
    body: formData
  });

  const result = await response.json();
  console.log(`Judgment: ${result.judgment}`);
  console.log(`Î”E: ${result.overall_delta_e.toFixed(2)}`);

  // ì‰í¬ ë¶„ì„
  const zoneBased = result.ink_analysis.zone_based.ink_count;
  const imageBased = result.ink_analysis.image_based.ink_count;
  console.log(`Zone-based: ${zoneBased}, Image-based: ${imageBased}`);

  return result;
}

// íŒŒë¼ë¯¸í„° ì¬ê³„ì‚°
async function recompute(imageId, sku, params) {
  const formData = new URLSearchParams();
  formData.append("image_id", imageId);
  formData.append("sku", sku);
  formData.append("params", JSON.stringify(params));

  const response = await fetch("http://localhost:8000/recompute", {
    method: "POST",
    headers: {"Content-Type": "application/x-www-form-urlencoded"},
    body: formData
  });

  return await response.json();
}

// ì‚¬ìš© ì˜ˆì‹œ
const fileInput = document.getElementById("fileInput");
const file = fileInput.files[0];

const result1 = await inspectImage(file, "SKU001");
const imageId = result1.image_id;

// íŒŒë¼ë¯¸í„° ì¡°ì • í›„ ì¬ê³„ì‚°
const result2 = await recompute(imageId, "SKU001", {
  smoothing_window: 20,
  min_gradient: 2.5
});
```

---

### 6.3 cURL

```bash
# ë‹¨ì¼ ì´ë¯¸ì§€ ê²€ì‚¬
curl -X POST "http://localhost:8000/inspect" \
  -F "file=@lens.jpg" \
  -F "sku=SKU001" \
  -F "run_judgment=true"

# íŒŒë¼ë¯¸í„° ì¬ê³„ì‚°
curl -X POST "http://localhost:8000/recompute" \
  -d "image_id=abc123" \
  -d "sku=SKU001" \
  -d 'params={"smoothing_window":20,"min_gradient":2.5}'

# ë°°ì¹˜ ê²€ì‚¬ (ZIP)
curl -X POST "http://localhost:8000/batch" \
  -F "zip_file=@batch.zip" \
  -F "sku=SKU001"

# ë¡œíŠ¸ ë¹„êµ
curl -X POST "http://localhost:8000/compare" \
  -F "reference_file=@golden.jpg" \
  -F "test_files=@lot1.jpg" \
  -F "test_files=@lot2.jpg" \
  -F "sku=SKU001"

# í—¬ìŠ¤ ì²´í¬
curl "http://localhost:8000/health"
```

---

## 7. ë¶€ë¡

### 7.1 Rate Limiting

**í˜„ì¬**: ì—†ìŒ
**í–¥í›„ ê³„íš**: ë¶„ë‹¹ 60 requests (ê°œë°œ í™˜ê²½), ë¶„ë‹¹ 600 requests (í”„ë¡œë•ì…˜)

### 7.2 CORS ì„¤ì •

**í˜„ì¬**: ëª¨ë“  origin í—ˆìš© (`allow_origins=["*"]`)
**í”„ë¡œë•ì…˜ ê¶Œì¥**:
```python
app.add_middleware(
    CORSMiddleware,
    allow_origins=["https://yourdomain.com"],
    allow_methods=["POST", "GET"],
    allow_headers=["*"]
)
```

### 7.3 íŒŒì¼ ì—…ë¡œë“œ ì œí•œ

- ìµœëŒ€ íŒŒì¼ í¬ê¸°: **50MB**
- ì§€ì› í˜•ì‹: **JPG, JPEG, PNG**
- ë™ì‹œ ì—…ë¡œë“œ: **ìµœëŒ€ 100ê°œ** (ë°°ì¹˜)

### 7.4 ìºì‹œ ì •ì±…

- ì´ë¯¸ì§€ ìºì‹œ (ì¬ê³„ì‚°ìš©): **TTL 15ë¶„**
- ë°°ì¹˜ ê²°ê³¼: **24ì‹œê°„ ë³´ê´€**

---

## 8. ë³€ê²½ ì´ë ¥

| ë‚ ì§œ | ë²„ì „ | ë³€ê²½ ë‚´ìš© |
|------|------|----------|
| 2025-12-13 | 0.1 | ì´ˆê¸° API (6ê°œ endpoints) |
| 2025-12-14 | 0.2 | `/recompute`, `/compare` ì¶”ê°€ (PHASE7) |
| 2025-12-14 | 0.3 | ì‰í¬ ë¶„ì„ ê²°ê³¼ í¬í•¨ |
| 2025-12-15 | 1.0 | ë¬¸ì„œ ì‘ì„± ì™„ë£Œ |

---

**ë¬¸ì˜**: í”„ë¡œì íŠ¸ ê°œë°œíŒ€ ë˜ëŠ” GitHub Issues
**ë§ˆì§€ë§‰ ê²€í† **: 2025-12-15
